text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,487 @@
+
+## Open Questions
+
+One of the goals for telemetry protocol is reducing CPU usage and memory pressure in garbage collected languages. These goals were not addressed as part of this RFC and remain open. One of the promising future ways to address this is finding a more CPU and memory efficient encoding mechanism.
+
+Another goal for telemetry protocol is achieving high compression ratios for telemetry data while keeping CPU consumption low. OTLP uses compression provided by gRPC transport. No further improvements to compression were considered as part of this RFC and are a future area of work.
+
+## Appendix A - Protocol Buffer Definitions
+
+This is Protocol Buffers schema for `Export` request and response:
+
+```
+// A request from client to server containing telemetry data to export.
+message ExportRequest {
+  // Unique sequential ID generated by the client.
+  uint64 id = 1;
+
+  // Telemetry data.
+  repeated NodeSpans nodeSpans = 2;
+  repeated NodeMetrics nodeMetrics = 3;
+}
+
+// A response to ExportRequest.
+message ExportResponse {
+  // ID of a response that the server acknowledges.
+  uint64 id = 1;
+
+  enum ResultCode {
+    // Telemetry data is successfully processed by the server.
+    Success = 0;
+    
+    // processing of telemetry data failed. The client MUST NOT retry 
+    // sending the same telemetry data. The telemetry data MUST be dropped. 
+    // This for example can happen when the request contains bad data and 
+    // cannot be deserialized or otherwise processed by the server.
+    FailedNoneRetryable = 1;
+    
+    // Processing of telemetry data failed. The client SHOULD record the 
+    // error and may retry exporting the same data after some time. This 
+    // for example can happen when the server is overloaded.
+    FailedRetryable = 2;
+  }
+  ResultCode result_code = 2;
+}
+
+// A list of spans from a Node.
+message NodeSpans {
+  Node node = 1;
+  Resource resource = 2;
+  repeated Span spans = 3;
+}
+
+// A list of metrics from a Node.
+message NodeMetrics {
+  Node node = 1;
+  Resource resource = 2;
+  repeated Metric metrics = 3;
+}
+```
+
+`Span`, `Metric`, `Node` and `Resource` schema definitions are defined in RFC (RFC number to be defined and linked from here).
+
+## Appendix B - Performance Benchmarks
+
+Benchmarking of OTLP vs other telemetry protocols was done using [reference implementation in Go](https://github.com/tigrannajaryan/exp-otelproto). In this OTLP implementation telemetry data, particularly Span batches are represented using a slightly modified Protocol Buffers schema taken from OpenCensus Protocol. Primary differences from OpenCensus schema are:
+
+- Eliminated TruncatableString and replaced by plain string.
+- Eliminated AttributesMap structure and used a map directly instead.
+- Replaced google.protobuf.Timestamp by int64 timestamp in Unix epoch nanoseconds.
```

Comment:
How much is this? I fell that using nanos since epoch is more like 2000s :)). Every library/language moved to the new 96bit representation of the Timestamp.

Reply:
I did not benchmark this change alone. I benchmarked the overall change. The difference from all changes is about 9% (compared to OpenCensus) in pure encoding/decoding time.

However, I think I should remove this from this RFC, it really does not belong here. It should be in the RFC which defines the protobufs schema of Span.

Reply:
Removed from this RFC.